MeesGame/Gameplay/Generator/MazeGeneratorMain.cs

```diff
@@ -130,5 +97,116 @@ private void MazeGenMain()
+            } else if(tiles.GetType(pointBetween.X, pointBetween.Y) == TileType.Floor)
+            {
+                behindDoorInfo[next.X, next.Y] = behindDoorInfo[current.X, current.Y];
+            }
+        }
+
+        private void AddTileInBetween()
+        {
+            // Default is a FloorTile
+            tileToAdd = new FloorTile();
+
+            // If a key has been placed, immediately place a door
+            if (doorsPlaced < keysPlaced)
+            {
+                tileToAdd = new DoorTile();
+                (tileToAdd as DoorTile).SecondarySpriteColor = ChooseColor(doorsPlaced);
+                doorsPlaced++;
+                // If there's a door between the tiles, a key is needed to pass, so the exit score should increase by more.
+                exitScore[next.X, next.Y] = exitScore[current.X, current.Y] + 20;
+            }
+            // Always have a chance to place a door after a key has been placed
+            else if (placeDoors && randomChance(doorChance))
+            {
+                tileToAdd = new DoorTile();
+                (tileToAdd as DoorTile).SecondarySpriteColor = ChooseColor(keysPlaced, true);
+                // If there's a door between the tiles, a key is needed to pass, so the exit score should increase by more.
+                exitScore[next.X, next.Y] = exitScore[current.X, current.Y] + 15;
+            }
+            tiles.Add(tileToAdd, pointBetween.X, pointBetween.Y);
+        }
+        
+        /// <summary>
+        /// Places a tile between the two points, creating the loop.
+        /// </summary>
+        private void AddLoopTile(Point one, Point two)
+        {
+            tileToAdd = new FloorTile();
+            behindDoorFlags difference = behindDoorInfo[one.X, one.Y] ^ behindDoorInfo[two.X, two.Y];
+            if (Enum.IsDefined(typeof(behindDoorFlags), difference))
+            {
+                tileToAdd = new DoorTile();
+                (tileToAdd as DoorTile).SecondarySpriteColor = ChooseColor((int)Math.Log((int)difference, 2));
+            }
+            if (difference != 0 || randomChance(loopHoleChance))
+            {
+                tileToAdd = new HoleTile();
+                if (randomChance(0.5))
+                    tileToAdd = new GuardTile();
+            }
+            tiles.Add(tileToAdd, PointBetween(one, two).X, PointBetween(one, two).Y);
+
+            // If a loop is added, the tiles are only behind the doors which both of them are behind.
+            // This should continue for further along the hallways of both loops, but doign that would require some form of flood-fill 
+            // which I don't have time for.
+            if(tileToAdd.TileType == TileType.Floor)
+            {
+                behindDoorInfo[one.X, one.Y] &= behindDoorInfo[two.X, two.Y];
```

Comment:
Floortiles are only chosen if the difference is 0, so this isn't needed. I have removed it.